### PR TITLE
Make present usable as a library

### PR DESCRIPTION
**Pull request by kmaasrud**
This (quite huge) PR makes present usable as a library. Along the way, I was able to remove quite a lot of unnecessary allocations (clones and collects), which means the resulting binary should be a bit faster. I have not benchmarked it though; planning on adding benchmarks in a future PR.

This refactor required a lot of restructuring of the internal API, but the CLI remains almost the same. The only modification I've done to the CLI is that I've change the `path` option to be an argument. This means that present is run like

```shell
present [OPTIONS] [PATH]
```

which I feel is much more sensible, and means we don't have to type `--path` all the time. As you can see, the path is optional - the default still being the current working drive.

## To do

- [x] Create working `lib.rs`
    - [x] Fix tests
- [x] Change `present --path <path>` to `present <path>`
- [x] Write docs
- [x] Bump version

### CODE REVIEW

**Comment by kmaasrud** [comment]
I everything should be good to go now! Just tell me if there is anything you want me to change 👍

File: README.md [hunk 1]
```diff
@@ -34,13 +34,15 @@ present 0.1.1
 Interpolate the standard output of arbitrary shell scripts into your markdown files
```
Comment (terror) [inline README.md:34]: I think we should add a library section in the readme as well, with a small example

File: src/arguments.rs [hunk 2]
```diff
@@ -1,20 +1,40 @@
-use crate::common::*;
+use crate::walker::{Walker, WalkerOptions};
+use present::{File, Result};
```
Comment (terror) [inline src/arguments.rs]: Small nitpick:

```suggestion
use {
  crate::walker::{Walker, WalkerOptions}, 
  present::{File, Result},
  clap::Parser as StructOpt
};
```

repeat for all files

File: README.md (hunk 1 above)
Comment (kmaasrud) [reply README.md:34]: Agreed!

File: src/arguments.rs (hunk 2 above)
Comment (kmaasrud) [reply src/arguments.rs]: Gotta keep things clean, I'm on it!

**Comment by kmaasrud** [comment]
I added the README to the doctests as well. Since the example snippet I included runs `present` on the README itself, all codeblocks get updated when running `cargo test`. How cool is that?!

Also, we probably want to update the video, but I'm guessing it's better if you do it :+1:

**Comment by terror** [comment]
> I added the README to the doctests as well. Since the example snippet I included runs `present` on the README itself, all codeblocks get updated when running `cargo test`. How cool is that?!
> 
> 
> 
> Also, we probably want to update the video, but I'm guessing it's better if you do it :+1:

Pretty neat! The rest looks good to me, just fix up the clippy lints on CI and it should be good to merge. 

I'll update the video in a separate PR.

Awesome work.

**Comment by kmaasrud** [comment]
> Pretty neat! The rest looks good to me, just fix up the clippy lints on CI and it should be good to merge.

Thanks! I'm away this weekend, but I'll get it done come monday 👌

**Comment by kmaasrud** [comment]
Should be good now!

**Pull request merged**